wi-fi/hal/cyhal_sdio.c

```diff
@@ -4,766 +4,128 @@
-			break;
-		case CYHAL_SDIO_CMD_SELECT_CARD:
-			cmd |= 1 << 20;   /* CICEN=1 */
-			cmd |= 1 << 19;   /* CCCEN=1 */
-			cmd |= 0x3 << 16; /* RSPTYP=3 */
-			break;
-		case CYHAL_SDIO_CMD_GO_INACTIVE_STATE:
-			/* no response */
-			break;
-		case CYHAL_SDIO_CMD_IO_RW_DIRECT:
-			cmd |= 1 << 20;   /* CICEN=1 */
-			cmd |= 1 << 19;   /* CCCEN=1 */
-			cmd |= 0x2 << 16; /* RSPTYP=2 */
-			break;
-		default:
-			return CYHAL_SDIO_RSLT_ERR_BAD_PARAM;
-	}
-
-	*(sdio_common.base + mix_ctrl) = 1U << 31;
-	*(sdio_common.base + blk_att) = 0;
-	*(sdio_common.base + cmd_arg) = argument;
-	*(sdio_common.base + cmd_xfer_typ) = cmd;
-
-	/* wait 1 ms max */
-	if (wait_for_cmd(0x1, 10, 100) < 0) {
-		reset_cmd_block();
-		return CYHAL_SDIO_RSLT_ERR_FUNC_RET(CYHAL_SDIO_RET_CMD_TIMEOUT);
-	}
-
-	/* clear status flags */
-	*(sdio_common.base + int_status) = 0x1; /* CC=1 */
-
-	/* retrieve response */
-	switch (command) {
-		case CYHAL_SDIO_CMD_GO_IDLE_STATE:
-			/* no response */
-			break;
-		case CYHAL_SDIO_CMD_SEND_RELATIVE_ADDR:
-			val = *(sdio_common.base + cmd_rsp0);
-			break;
-		case CYHAL_SDIO_CMD_IO_SEND_OP_COND:
-			val = *(sdio_common.base + cmd_rsp0);
-			break;
-		case CYHAL_SDIO_CMD_SELECT_CARD:
-			val = *(sdio_common.base + cmd_rsp3);
-			break;
-		case CYHAL_SDIO_CMD_GO_INACTIVE_STATE:
-			/* no response */
-			break;
-		case CYHAL_SDIO_CMD_IO_RW_DIRECT:
-			val = *(sdio_common.base + cmd_rsp0);
-			break;
-		default:
-			return CYHAL_SDIO_RSLT_ERR_BAD_PARAM;
-	}
-
-	if (response)
-		*response = val;
-
-	return CY_RSLT_SUCCESS;
+	/* block size is 0 for some reason, hardcode instead */
+	return sdio_config(config->frequencyhal_hz, 64);
 }
 
 
-/* NOTE: obj is ignored - state is kept in sdio_common */
+/* NOTE: obj is ignored, state instance is held by the SDIO API*/
 cy_rslt_t cyhal_sdio_send_cmd(const cyhal_sdio_t *obj, cyhal_transfer_t direction,
 	cyhal_sdio_command_t command, uint32_t argument, uint32_t *response)
 {
-	cy_rslt_t res;
+	int rslt;
+	uint8_t data;
 
-	// cy_log_msg(CYLF_SDIO, CY_LOG_DEBUG, "cyhal_sdio_send_cmd (dir=%u cmd=%u arg=%x)\n", direction, command, argument);
+	sdio_cmd_arg_t arg;
+	arg.uint = argument;
 
-	for (unsigned int i = 1; i <= 10; ++i) {
-		mutexLock(sdio_common.cmd_lock);
-		res = cyhal_sdio_send_cmd_internal(direction, command, argument, response);
-		mutexUnlock(sdio_common.cmd_lock);
-
-		if (res == CY_RSLT_SUCCESS)
-			break;
-
-		cy_log_msg(CYLF_SDIO, CY_LOG_WARNING, "repeating SDIO cmd (cnt=%u)\n", i);
-
-		usleep(1000);
-	}
-
-	return res;
-}
-
-
-static cy_rslt_t cyhal_sdio_bulk_transfer_internal(cyhal_transfer_t direction, uint32_t argument, uint32_t *data, uint16_t length, uint32_t *response)
-{
-	uint32_t val, cmd, mix, blk, block_mode, count;
-
-	block_mode = !!(argument & (1 << 27));
-	count = argument & 0x1ff;
-
-	if (response)
+	/* WHD only uses send cmd api call for
+	 * transfers after init commands which are
+	 * issued by sdio_init function, hence
+	 * just indicate success and exit */
```

Comment:
I'm afraid we can't do it this way - assuming a specifiic behaviour of the upper layer, in this case the whd driver. It might change in the future, we need to keep it generic still. Unfortunately I see no other option here than exporting `sdio_sendCmd`, which is now static, and maybe check if the cmd is CYHAL_SDIO_CMD_IO_RW_DIRECT and use transferDirect function in this case and sendCmd otherwise.